site/migration-guides/entrypoint-v7.md

```diff
@@ -0,0 +1,231 @@
+//   factory?: string;  [!code ++]
+//   factoryData?: BytesLike;  [!code ++]
+//   callData: BytesLike;
+//   callGasLimit?: BigNumberish;
+//   verificationGasLimit?: BigNumberish;
+//   preVerificationGas?: BigNumberish;
+//   maxFeePerGas?: BigNumberish;
+//   maxPriorityFeePerGas?: BigNumberish;
+//   paymasterAndData: BytesLike | "0x";  [!code --]
+//   paymaster?: string;  [!code ++]
+//   paymasterVerificationGasLimit?: BigNumberish;  [!code ++]
+//   paymasterPostOpGasLimit?: BigNumberish;  [!code ++]
+//   paymasterData?: BytesLike;  [!code ++]
+//   signature: BytesLike;
+// }
+```
+
+### [`UserOperationOverrides`](/resources/types#useroperationoverrides)
+
+Similar to the updated `UserOperation` type from `EntryPoint` v0.6 to v0.7, `UserOperationOverrides` types are different from the two version operations.
+
+#### `paymasterAndData` field replaced with `paymaster` address and the `paymasterData` separated in addition to the 2 added `paymaster` operation gas limit fields
+
+```ts
+export type UserOperationOverrides<'0.7.0'> = {
+  ...
+  paymasterAndData?: EmptyHex;  // [!code --]
+  paymaster?: string;  // [!code ++]
+  paymasterData?: EmptyHex;  // [!code ++]
+  paymasterVerificationGasLimit?: BigNumberish | Multiplier;  // [!code ++]
+  paymasterPostOpGasLimit?: BigNumberish | Multiplier;  // [!code ++]
+}
+```
+
+::: tip Note
+Note that per [`ERC-4337`](https://eips.ethereum.org/EIPS/eip-4337#parameters), paymaster fields (`paymaster`, `paymasterData`, `paymasterValidationGasLimit`, `paymasterPostOpGasLimit`) either all exist, or none in the final `UserOperation` request sent to the connected Bundler.
+:::
+
+#### `stateOverride` field added to `UserOperationOverrides` for validating and estimating gas for transactions
+
+The state overrides during transaction calls (e.g., [`eth_call`](https://docs.alchemy.com/reference/eth-call) and [`eth_estimateGas`](https://docs.alchemy.com/reference/eth-estimategas)) allow for more flexible testing and gas estimation. However, not all networks support this feature. For networks lacking state overrides, `EntryPoint` v0.7 introduced the `delegateAndRevert()` function to be used as a workaround. It aids in validating and estimating gas for transactions by simulating conditions without permanently changing the on-chain state.
+
+Although some bundlers, such as Alchemy [`rundler`](https://github.com/alchemyplatform/rundler), does support the `stateOverride` feature for user operation gas estimation for both EntryPoint v0.6 and v0.7 user operations, until Entrypoint v0.6, the [`eth_estimateUserOperationGas`](https://eips.ethereum.org/EIPS/eip-4337#-eth_estimateuseroperationgas) RPC method did not include the optional `stateOverride` parameter per specification. From EntryPoint v0.7, however, `stateOverride` has been added to officially added as an optional parameter to the gas estimation method, thanks to the removal of network requirement to support state override feature.
```

Comment:
Typo: "does support the `stateOverride`" -> "do support the `stateOverride`"

Also,
> thanks to the removal of network requirement to support state override feature
is incorrect. `delegateAndRevert` is not useful for implementing the `stateOverride` field of bundler methods. It's intended use is internal to bundlers, and it does not allow general state overrides that would be necessary to support the `stateOverride` parameter.

Reply:
@dphilipson this note is directly from official 0.7 release note:
<img width="1073" alt="Screenshot 2024-04-23 at 3 01 06 PM" src="https://github.com/alchemyplatform/aa-sdk/assets/3278577/51eca104-92c9-4fab-885a-9511228d18bb">

Reply:
Updated as "The state overrides during transaction calls (e.g., [`eth_call`](https://docs.alchemy.com/reference/eth-call) and [`eth_estimateGas`](https://docs.alchemy.com/reference/eth-estimategas)) allow for more flexible testing and gas estimation. However, not all networks support this feature. For networks lacking state overrides, `EntryPoint` v0.7 introduced the `delegateAndRevert()` helper in the `EntryPoint` contract to accommodate nodes without state overrides in `eth_call` and `eth_estimateGas`. It aids in validating and estimating gas for transactions by simulating conditions without permanently changing the on-chain state."

Reply:
Yeah, the point that was incorrect in the original wording was that it implied that `delegateAndRevert` supported the **state override parameter** of `eth_estimateUserOperationGas`. The wording in 4337's note doesn't say this: it says that `delegateAndRevert` is used to support the implementation of validation and gas estimation, which is true.

Reply:
Actually, I would say the current wording still has the incorrect implication that `delegateAndRevert` allows the state overrides parameter on networks that don't support it (really: it's less network-support and more node client-support). It's in the section titled "`stateOverride` field added to `UserOperationOverrides` for validating and estimating gas for transactions".

I might just take out mention of `delegateAndRevert` in this section. It's not related to the `stateOverride` parameter.

Reply:
ok

Reply:
Good question. Here's the full explaination as I'm thinking of it.

One other change in the entry point from 0.6 to 0.7 is that the 0.6 entry point used to contain some functions which were intended to be called by bundlers purely for simulation's sake: `simulateValidation` and `simulateHandleOp`.

In 0.7, they decided that there was no need to deploy functions that would only ever be used off-chain, so instead they defined a new contract, `EntryPointSimulations`, which extends `EntryPoint` with the simulation functions. `EntryPointSimulations` is not deployed on-chain; instead, bundlers are now expected to use state overrides in their internal calls to `eth_call`/`eth_estimateGas`. The bundler will still call `simulateValidation` or `simulateHandleOp` on the entry point address as before, but since the real entry point doesn't have these methods anymore, the bundler needs to spoof the state of the blockchain to replace the entry point code with the version that has these methods.

Bundlers can do this using state overrides, but the 4337 authors didn't want to _assume_ that state overrides are available. This may be because state overrides are not part of the standard eth api, but rather an extension added by Geth (and copied by other node implementations that want to be compatible). Notice how state overrides don't appear in the [official eth jsonrpc spec for `eth_call`](https://ethereum.github.io/execution-apis/api-documentation/).

So the entry point provides an escape hatch for node clients that don't support state overrides, `delegateAndRevert`, which effectively allows the caller to do the same thing by "overriding" the entry point's code with the entry point simulation code. `delegateAndRevert` uses delegatecall internally, and in general a delegatecall can be thought of as replacing a contract's code with a different contract for the duration of a single call.

`delegateAndRevert` is a lot less powerful than full state overrides. `delegateAndRevert` only allows the caller to replace the code of the entry point contract, while state overrides allow the caller to replace the code, storage, balance, or nonce of any address. But for the purpose of calling the `EntryPointSimulation` functions, just replacing the entry point code is enough.

Note that this is all totally separate from the optional state overrides parameter to `eth_estimateUserOperationGas`. That parameter is giving the caller the full set of override options as you would get when using overrides in an `eth_call`, which as stated above is much more powerful than what `delegateAndRevert` allows.